mapstore.go

```diff
@@ -20,38 +22,64 @@ func (e *InvalidKeyError) Error() string {
 	return fmt.Sprintf("invalid key: %x", e.Key)
 }
 
+// ErrWrongKeySize is returned when a key has a different size than the key size.
+var ErrWrongKeySize = errors.New("wrong key size")
+
 // SimpleMap is a simple in-memory map.
 type SimpleMap struct {
-	m map[string][]byte
+	m       map[string][]byte
+	keySize int
 }
 
 // NewSimpleMap creates a new empty SimpleMap.
-func NewSimpleMap() *SimpleMap {
+func NewSimpleMap(keySize int) *SimpleMap {
 	return &SimpleMap{
-		m: make(map[string][]byte),
+		m:       make(map[string][]byte),
+		keySize: keySize,
 	}
 }
 
 // Get gets the value for a key.
 func (sm *SimpleMap) Get(key []byte) ([]byte, error) {
+	if err := sm.checkKeySize(key); err != nil {
+		return nil, err
+	}
 	if value, ok := sm.m[string(key)]; ok {
 		return value, nil
 	}
 	return nil, &InvalidKeyError{Key: key}
 }
 
+// GetKeySize gets the key size of the map store.
+func (sm *SimpleMap) GetKeySize() int {
+	return sm.keySize
+}
+
 // Set updates the value for a key.
 func (sm *SimpleMap) Set(key []byte, value []byte) error {
+	if err := sm.checkKeySize(key); err != nil {
+		return err
+	}
 	sm.m[string(key)] = value
 	return nil
 }
 
 // Delete deletes a key.
 func (sm *SimpleMap) Delete(key []byte) error {
+	if err := sm.checkKeySize(key); err != nil {
+		return err
+	}
 	_, ok := sm.m[string(key)]
 	if ok {
 		delete(sm.m, string(key))
 		return nil
 	}
 	return &InvalidKeyError{Key: key}
 }
+
+func (sm *SimpleMap) checkKeySize(key []byte) error {
```

Comment:
I think the SMT is by construction assumed to be perfect. Allowing different key sizes...doesn't really make sense? How would that even work?

Reply:
I thought it should only support a fixed size of keys, no ? 
The `checkKeySize` is not exported (If I am not wrong), it is only used internaly. What it does is verify the length of the provided key if it matches the key size or return an error otherwise. I put it in a function to avoid rewriting the same code over and over again. 

Do you mean we can have any key length ? I guess that wasn't the case even before this PR, since it was using a specific hashing algorithm to hash the keys and the depth of the tree was gotten via `hasher.size()`...

Reply:
I totally agree, that SMT require fixed key size internally. My only concern is that, previously our SMT public interface was a black-box/high-level/general-purpose-container type of thing (see README.md with `_, _ = tree.Update([]byte("foo"), []byte("bar"))`). This change is a complete pivot, where de-facto part of implementation is exposed to the user, and user has to be aware of this.

Reply:
true, it is a breaking change. However, it will impact the interface only at the time of creation of the map stores and verification of proofs. 
For `addBranch`, it was a mistake from my side to put the `keySize` as a parameter since it should be checked against the `dsmst`'s map store and not against the one passed in params (will make the change shortly).

Previously, I think when verifying a proof, we were passing a `hasher`, from which we extracted the `keySize`. So, we were verifying a proof against a certain `hasher` and a `keySize`. Except that the latter wasn't exposed since it was extracted from the hasher. Now that the key size doesn't relate to the hasher, it's alright to pass it as parameter.

What do you think ?

Reply:
The actual function signatures of `Get`, `Set`, `Update`, `Delete`, etc are not changed, but there is an extra "contract" added (fixed length keys). While updating SMT dependency, user has to potentially update all invocations of those methods.

Reply:
Aaah, got you now :+1: makes sense